WordPress/gutenberg
# Block Editor: Update Block Editor classNames to convention

Previously: #14112

This nightmare of a pull request seeks to update class names for components moved from the `editor` package to `block-editor`, in order to continue respecting the [CSS class naming guidelines](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#naming) which prescribe that each component be prefixed by its package directory. It does so without breaking compatibility by _keeping_ the `.editor-` prefix as a compatibility mechanism.

In its current incarnation, it is as conservative as possible to keeping full compatibility. Unlike my [previous comment](https://github.com/WordPress/gutenberg/pull/14112#issuecomment-471575882) by which I'd jynxed myself, automated transforms are highly fraught with issues (dealing with `classnames`, string template interpolation, etc). Conversely, I chose not to get into specifics of what's "worth" keeping for compatibility, for the sake of keeping the effort short-lived and as non-contentious as possible. In both of these instances, I think revisiting in the future may be appropriate.

There is an argument to be made that we make a specific exception to the CSS naming guidelines for the `block-editor` module, allowing both `editor` and `block-editor` components to share the `editor-` prefix for the sake of avoiding this tedious transition. My cautions against this are that (a) exceptions are bad because they break expectations and (b) it introduces the risk of unintentional conflict when working with components across the two otherwise distinct modules.

**Implementation notes:**

The changes here were partly RegEx-assisted mass string replacements, with a supplementary (sometimes involved) manual curation.

Specific replacement patterns:

1. Replace string occurrences, exempting [icon slugs](https://github.com/WordPress/dashicons/tree/master/sources/svg):
   - Replace: `(['"])editor-(?!aligncenter|alignleft|alignright|bold|break|code|contract|customchar|expand|help|indent|insertmore|italic|justify|kitchensink|ltr|ol-rtl|ol|outdent|paragraph|paste-text|paste-word|quote|removeformatting|rtl|spellcheck|strikethrough|table|textcolor|ul|underline|unlink|video)([^ '"]+)`
   - With: `$1editor-$2 block-editor-$2`
   - **Note:** In initial passes, this missed some components where the string does not _begin_ with `editor-` ([example](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/block-editor/src/components/block-list/block.js#L434))
2. Remove `editor-` occurrence for functions expecting a singular argument
   - Replace: `(hasClass|contains)\( 'editor-.*? `
   - With: `$1( '`
3. Remove `editor-` occurrence for IDs
   - Replace: `(id|htmlFor|aria-.+?)(="|': ')editor-.*? `
   - With: `$1$2`
   - **Note:** There is no backwards-compatibility for IDs. They are largely used exclusively for the in-app label association.
4. Replace CSS selector occurrences for [deprecated components](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/editor/src/components/deprecated.js)
   - Replace: `\.editor-(autocomplete|alignment-toolbar|block-alignment-toolbar|block-controls|block-edit|block-format-controls|block-icon|block-inspector|block-list|block-mover|block-navigation-dropdown|block-selection-clearer|block-settings-menu|block-title|block-toolbar|color-palette|contrast-checker|copy-handler|default-block-appender|inserter|inner-blocks|inspector-advanced-controls|inspector-controls|panel-color-settings|plain-text|rich-text|media-placeholder|media-upload|multi-blocks-switcher|multi-select-scroll-into-view|navigable-toolbar|observe-typing|preserve-scroll-in-reorder|skip-to-selected-block|url-input|url-popover|warning|writing-flow)`
   - With: `.block-editor-$1`
   - **Note:** This does not cover all migrated components, only those which had been exported publicly. Additional searches for `.editor-` were manually performed (guided in some cases by "post" exception `\.editor-(?!post-)`).

Notable exceptions:

- We have some occurrences of [_clever_](https://softwareengineering.stackexchange.com/questions/25276/why-is-cleverness-considered-harmful-in-programming-by-some-people) concatenation ([example](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/block-editor/src/components/panel-color-settings/index.js#L94)). Template string interpolation in general proved tricky.
- Native components bypassed compatibility and upgraded the class names directly ([example](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/block-editor/src/components/plain-text/index.native.js#L44))

**Testing Instructions:**

Give it a thorough shake, paying very close attention to visual regressions.

End-to-end tests should also reveal some issues, if present.

```
npm run test-e2e
```

Most useful review will be in considering what may have been overlooked or wrongly revised.
